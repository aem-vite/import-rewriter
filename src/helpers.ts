import { createHash } from 'crypto'
import _debug from 'debug'
import { existsSync, readFileSync } from 'fs'
import { join } from 'path'

import type { InputOptions, NormalizedOutputOptions, OutputAsset, OutputChunk } from 'rollup'

import type { AEMLongCacheConfiguration, BundlesImportRewriterOptions } from './types'

const entryPaths = new Set<string>()

export const debug = _debug('aem-vite-import-rewriter')
export const relativePathPattern = /([.]{1,2}\/)+/

/**
 * Retrieve the entry file paths.
 *
 * @returns entry paths
 */
export function getEntryPaths() {
  return entryPaths
}

/**
 * Add the provided `path` to the available entry paths.
 *
 * @param path the path to the entry file
 */
export function setEntryPath(path: string): void {
  entryPaths.add(path)
}

/**
 * Determine if the provided `assetOrChunk` object is a chunk instance.
 *
 * @param assetOrChunk an instance of either `OutputAsset` or `OutputChunk`
 * @returns `true` when `OutputChunk`, otherwise `false`
 */
export function isOutputChunk(assetOrChunk: OutputAsset | OutputChunk): assetOrChunk is OutputChunk {
  return typeof (assetOrChunk as OutputChunk).imports !== undefined
}

/**
 * Generate an MD5 checksum for the provided `source` input.
 *
 * @param source raw input
 * @returns an MD5 checksum of the `source` input
 */
export function generateChecksum(source: string): string {
  return createHash('md5').update(source, 'utf8').digest('hex')
}

/**
 * Generates a unique cache key for the provided `path` that importing modules will use to
 * load the correct entry file.
 *
 * @param entryPath the base ClientLib path
 * @param keyFormat HTML Library Manager long cache format
 * @returns a unqiue cache key for the provided `path`
 */
export function getCacheKey(entryPath: string, keyFormat: AEMLongCacheConfiguration['keyFormat']): string {
  let keyFormatString = ''

  switch (keyFormat) {
    case 'cloud':
      keyFormatString = 'lc-%s-lc.%m'
      break
    case 'acs-classic':
      keyFormatString = '%s.%m'
      break
    case 'acs-modern':
      keyFormatString = '%m.ACSHASH%s'
      break
  }

  const combinedContents = [...entryPaths].map((entry) => {
    const path = join(entryPath, entry)

    return existsSync(path) ? readFileSync(path).toString() : ''
  })

  return keyFormatString.replace('%s', generateChecksum(combinedContents.join('')))
}

/**
 * Constructs an AEM ClientLib path for the current configuration.
 *
 * @param options import rewriter options
 * @param forImport is the ClientLib path for an import?
 * @param withChecksum should a checksum be generated?
 * @param rollupOptions rollup options object
 * @returns the real AEM ClientLib path
 */
export function getAemClientLibPath(
  options: BundlesImportRewriterOptions,
  forImport = false,
  withChecksum = false,
  rollupOptions?: NormalizedOutputOptions,
): string {
  let path = options.publicPath

  if (forImport) {
    return `${path}/${options.resourcesPath}/`
  }

  if (withChecksum && options.caching && options.caching.enabled && rollupOptions !== undefined) {
    const entryPath = rollupOptions.dir as string

    // Append the MD5 checksum and minified extension to the path
    path = `${path}.${getCacheKey(entryPath, options.caching.keyFormat)}`
    path = path.replace('.%m', options.minify === true ? '.min' : '')
  }

  return `${path}.js`
}

/**
 * Converts the standard import path into an AEM compliant ClientLib path.
 *
 * @param parentPath path of the parent (chunk)
 * @param path the import path
 * @param options import rewriter options
 * @param entryAliases rollup entry aliases
 * @returns an AEM compliant ClientLib path
 */
export function getReplacementPath(
  parentPath: string,
  path: string,
  options: BundlesImportRewriterOptions,
  entryAliases: NonNullable<InputOptions['input']>,
): string {
  const isEntryPath = entryPaths.has(parentPath)

  if (isEntryPath) {
    return path.replace(new RegExp(`^${relativePathPattern.source}`), getAemClientLibPath(options, true))
  }

  return isInputAnEntryAlias(path, entryAliases)
    ? path.replace(
        new RegExp(`${relativePathPattern.source}${path.replace(relativePathPattern, '')}`),
        getAemClientLibPath(options),
      )
    : path
}

/**
 * Determines if the provided `input` is an entry alias or not.
 *
 * @param input the string to check against
 * @param entryAliases rollup entry aliases
 * @returns `true` when `input` is an alias, otherwise `false`
 */
export function isInputAnEntryAlias(input: string, entryAliases: NonNullable<InputOptions['input']>) {
  const entryAliasesExpr = new RegExp(`^[./]+(${Object.keys(entryAliases).join('|')})\\.js$`)

  return input.match(entryAliasesExpr)?.[0] ? true : false
}
