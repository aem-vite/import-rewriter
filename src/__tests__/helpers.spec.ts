import { beforeEach, afterEach, expect, test, vi } from 'vitest'

import { generateChecksum, getAemClientLibPath, getCacheKey, getEntryPaths, setEntryPath } from '../helpers'

import type { NormalizedOutputOptions } from 'rollup'

import type { BundlesImportRewriterOptions } from '../types'

const PUBLIC_PATH = '/etc.clientlibs/aem-vite/clientlibs/clientlib-site'

const checksum = generateChecksum('test')

const defaultPluginOptions: BundlesImportRewriterOptions = {
  publicPath: PUBLIC_PATH,
  resourcesPath: 'resources/js',
}

const rollupOptions = {
  dir: '/testing/aem-vite',
} as NormalizedOutputOptions

beforeEach(() => {
  vi.mock('node:fs', async () => {
    return {
      ...(await vi.importActual<typeof import('node:fs')>('node:fs')),
      existsSync: vi.fn().mockReturnValue(true),
      readFileSync: vi.fn().mockReturnValue('test'),
    }
  })
})

afterEach(() => {
  vi.clearAllMocks()
})

test('entry path should be set', () => {
  setEntryPath('importer')

  expect(getEntryPaths()).toStrictEqual(new Set(['importer']))
})

test('checksum should match', () => {
  expect(checksum).toEqual('098f6bcd4621d373cade4e832627b4f6')
  expect(checksum).toMatchSnapshot()
})

test('cache key should match', () => {
  const checksum = generateChecksum('test')
  const cacheKey = getCacheKey('test', 'cloud')
  const classicAcsCacheKey = getCacheKey('test', 'acs-classic')
  const modernAcsCacheKey = getCacheKey('test', 'acs-modern')

  expect(cacheKey).toEqual(`lc-${checksum}-lc.%m`)
  expect(classicAcsCacheKey).toEqual(`${checksum}.%m`)
  expect(modernAcsCacheKey).toEqual(`%m.ACSHASH${checksum}`)

  expect([cacheKey, classicAcsCacheKey, modernAcsCacheKey, checksum]).toMatchSnapshot()
})

test("invalid cache key should default to 'cloud'", () => {
  expect(getCacheKey('test', undefined)).toStrictEqual('lc-098f6bcd4621d373cade4e832627b4f6-lc.%m')
})

test('clientlib path is correct', () => {
  const path = getAemClientLibPath({ ...defaultPluginOptions }, false, false)

  expect(path).toEqual(`${PUBLIC_PATH}.js`)
})

test('clientlib path should contain hash', async () => {
  const options: BundlesImportRewriterOptions = {
    ...defaultPluginOptions,

    caching: {
      enabled: true,
      keyFormat: 'cloud',
    },
  }

  const { existsSync, readFileSync } = await import('node:fs')

  const path = getAemClientLibPath(options, false, true, rollupOptions)
  const minifyPath = getAemClientLibPath({ ...options, minify: true }, false, true, rollupOptions)

  expect(path).toEqual(`${PUBLIC_PATH}.lc-${checksum}-lc.js`)
  expect(minifyPath).toEqual(`${PUBLIC_PATH}.lc-${checksum}-lc.min.js`)

  expect(existsSync).toHaveBeenCalledTimes(2)
  expect(readFileSync).toHaveBeenCalledTimes(2)
})

test('clientlib path should contain invalid hash', async () => {
  const options: BundlesImportRewriterOptions = {
    ...defaultPluginOptions,

    caching: {
      enabled: true,
      keyFormat: 'acs-modern',
    },
  }

  vi.resetAllMocks()
  vi.doMock('node:fs', async () => {
    return {
      ...(await vi.importActual<typeof import('node:fs')>('node:fs')),
      existsSync: vi.fn().mockReturnValue(false),
    }
  })

  const checksum = generateChecksum('')
  const path = getAemClientLibPath(options, false, true, rollupOptions)

  expect(path).toEqual(`${PUBLIC_PATH}.ACSHASH${checksum}.js`)
  expect(path).toMatchSnapshot()
})

test('clientlib resource path should match', async () => {
  const options: BundlesImportRewriterOptions = {
    ...defaultPluginOptions,

    caching: {
      enabled: true,
      keyFormat: 'cloud',
    },
  }

  const path = getAemClientLibPath(options, true, false, rollupOptions)

  expect(path).toEqual(`${PUBLIC_PATH}/${options.resourcesPath}/`)
})
