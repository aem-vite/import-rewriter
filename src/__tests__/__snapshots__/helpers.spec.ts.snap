// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cache key should match 1`] = `
[
  "lc-098f6bcd4621d373cade4e832627b4f6-lc.%m",
  "098f6bcd4621d373cade4e832627b4f6.%m",
  "%m.ACSHASH098f6bcd4621d373cade4e832627b4f6",
  "098f6bcd4621d373cade4e832627b4f6",
]
`;

exports[`checksum should match 1`] = `"098f6bcd4621d373cade4e832627b4f6"`;

exports[`clientlib path should contain invalid hash 1`] = `"/etc.clientlibs/aem-vite/clientlibs/clientlib-site.ACSHASHd41d8cd98f00b204e9800998ecf8427e.js"`;
